felix/bpf/proxy/proxy_test.go

```diff
@@ -536,6 +537,133 @@ var _ = Describe("BPF Proxy", func() {
+				TypeMeta: typeMetaV1("Service"),
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "customer",
+					Namespace: "yaobank",
+					Labels: map[string]string{
+						"app": "customer",
+					},
+					Annotations: map[string]string{
+						v1.AnnotationTopologyAwareHints: "Auto",
+					},
+				},
+				Spec: v1.ServiceSpec{
+					ClusterIP: "10.96.98.148",
+					Ports: []v1.ServicePort{
+						{
+							Name:       "http",
+							NodePort:   30180,
+							Port:       80,
+							Protocol:   v1.ProtocolTCP,
+							TargetPort: intstr.FromInt(80),
+						},
+					},
+					Type: v1.ServiceTypeNodePort,
+					Selector: map[string]string{
+						"app": "customer",
+					},
+				},
+			}
+
+			testTopologyEndpoint := &v1.Endpoints{
+				TypeMeta: typeMetaV1("Endpoints"),
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "customer",
+					Namespace: "yaobank",
+					Labels: map[string]string{
+						"app": "customer",
+					},
+				},
+				Subsets: []v1.EndpointSubset{
+					{
+						Addresses: []v1.EndpointAddress{
+							{
+								IP:       "192.168.152.3",
+								NodeName: strPtr("stevepro-bz-wyo0-worker2"),
+								TargetRef: &v1.ObjectReference{
+									Kind:      "Pod",
+									Name:      "customer-5dd6b8478-k9l5l",
+									Namespace: "yaobank",
+								},
+							},
+							{
+								IP:       "192.168.185.67",
+								NodeName: strPtr("stevepro-bz-wyo0-worker3"),
+								TargetRef: &v1.ObjectReference{
+									Kind:      "Pod",
+									Name:      "customer-5dd6b8478-dswzk",
+									Namespace: "yaobank",
+								},
+							},
+						},
+						Ports: []v1.EndpointPort{
+							{
+								Name:     "http",
+								Port:     80,
+								Protocol: v1.ProtocolTCP,
+							},
+						},
+					},
+				},
+			}
+
+			zones := []string{"us-west-2a", "us-west-2b"}
+			testTopologySvcEpsSlice := epsToSliceWithZones(testTopologyEndpoint, zones)
+
+			k8s = fake.NewSimpleClientset(testTopologySvc, testTopologySvcEpsSlice)
+
+			BeforeEach(func() {
+				By("creating proxy with fake client and mock syncer", func() {
+					var err error
+
+					syncStop = make(chan struct{})
+					dp = newMockSyncer(syncStop)
+
+					opts := []proxy.Option{proxy.WithImmediateSync()}
+
+					p, err = proxy.New(k8s, dp, "testnode", opts...)
+					Expect(err).NotTo(HaveOccurred())
+				})
+			})
+
+			AfterEach(func() {
+				By("stopping the proxy", func() {
+					close(syncStop)
+					p.Stop()
+				})
+			})
+
+			It("should invoke k8s CategorizeEndpoints", func() {
+
+				By("successfully syncing the service and endpoint maps", func() {
+					dp.checkState(func(s proxy.DPSyncerState) {
+						Expect(len(s.SvcMap)).To(Equal(1))
+						Expect(len(s.EpsMap)).To(Equal(1))
+
+						// Test Topology Aware Hints function.
+						nodeLabels := map[string]string{v1.LabelTopologyZone: "us-west-2a"}
+						clusterEndpoints, localEndpoints, allReachableEndpoints, hasAnyEndpoints := proxy.InvokeCategorizeEndpoints(s.SvcMap, s.EpsMap, nodeLabels)
```

Comment:
this function should not be called here but somewhere in proxy.go This dp.checkState is only supposed to check that the state we are passing to syncer it what we expect. So maybe that rewrite above is not necessary is you call it from the proxy.go code.

Reply:
The `InvokeCategorizeEndpoints` above is just a helper function - the real call to K8s `CategorizeEndpoints` function is invoked in proxy.go  
PS: ideally I would have liked to have created the svcMap and epsMap objects directly and invoke in isolation i.e. unit tested but I see that the hintsAnnotation, which must be set on BaseServiceInfo object in order for Topology Aware Hints to work, is private.  I did see that if I create a service and endpoint the syncer would "generate" the corresponding svcMap and epsMap resources that could be used to test Topology Aware Hints but the current code here does feel wrong